app/assets/javascripts/components/flexible_roster.jsx

```diff
@@ -0,0 +1,158 @@
+import SortHelpers from '../helpers/sort_helpers.jsx';
+
+window.shared || (window.shared = {});
+
+export default React.createClass({
+  displayName: 'FlexibleRoster',
+
+  propTypes: {
+    rows: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
+    columns: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
+    initialSortIndex: React.PropTypes.number
+  },
+
+  getInitialState () {
+    return {
+      sortByIndex: this.props.initialSortIndex,
+      sortDesc: true
+    };
+  },
+
+  orderedRows () {
+    const sortedRows = this.sortedRows();
+
+    if (!this.state.sortDesc) return sortedRows.reverse();
+
+    return sortedRows;
+  },
+  
+  sortedRows () {
+    const rows = this.props.rows;
+    const columns = this.props.columns;
+    const sortByIndex = this.state.sortByIndex;
+    const key = columns[sortByIndex].key;
+
+    if ('sortFunc' in columns[sortByIndex]) {
+      return rows.sort((a,b) => columns[sortByIndex].sortFunc(a,b,key));
+    }
+    else {
+      return rows.sort((a, b) => SortHelpers.sortByString(a, b, key));
+    }
+    
+  },
+
+  headerClassName (sortByIndex) {
+    // Using tablesort classes here for the cute CSS carets,
+    // not for the acutal table sorting JS (that logic is handled by this class).
+
+    if (sortByIndex !== this.state.sortByIndex) return 'sort-header';
+
+    if (this.state.sortDesc) return 'sort-header sort-down';
+
+    return 'sort-header sort-up';
+  },
+
+  onClickHeader(sortByIndex) {
+    if (sortByIndex === this.state.sortByIndex) {
+      this.setState({ sortDesc: !this.state.sortDesc });
+    } else {
+      this.setState({ sortByIndex: sortByIndex});
+    }
+  },
+  
+  render () {
+    return (
+      <div className='FlexibleRoster'>
+        <table id='roster-table' className='roster-table' style={{ width: '100%' }}>
+          <thead>
+            {this.renderSuperHeaders()}
+            {this.renderHeaders()}
+          </thead>
+          {this.renderBody()}
+        </table>
+      </div>
+    );
+  },
+
+  renderSuperHeaders() {
+    var currentCount = 0;
+    const columns = this.props.columns;
+
+
+    var superHeaders = [];
+
+    for (var i=0; i<columns.length; i++) {
```

Comment:
This `for` loop is hard to read, as is a some of the code within it:

```
if(i+1 == columns.length || columns[i].group != columns[i+1].group) 
```

I would suggest you refactor using lodash, a helper library for these sorts of things that we use throughout the JS to make complex functions simpler and easy to read: https://lodash.com/docs/4.17.4

Reply:
@alexsoble lodash is cool, not sure how it would help in this instance though since I'm trying to do something pretty specific. Essentially, I'm collapsing the structure down to the number of sequentially repeated elements. 

I thought about adding more structure / hierarchy to the configuration (columns) but it doesn't seem worth it.

See if this code is more readable:

for (let i=0; i<columns.length; i++) {
      // group for the current column
      let itemGroup = columns[i].group;

      // group for the next column for comparison
      // set to null if this is the last column
      let nextItemGroup = columns.length > i+1 ? columns[i+1].group : null;

      // count of items with the same group
      // increment in the beginning since colSpan starts at 1
      currentCount++;
      
      // if the current item doesn't equal the next
      // push the super header with a length of currentCount
      // and reset currentCount for a new column group
      if(itemGroup != nextItemGroup) {
        superHeaders.push({label: itemGroup, span: currentCount});
        currentCount = 0;
      }
    }

Reply:
@jhilde Yes, that's helpful! Having a `nextItemGroup` variable, plus a simpler inequality comparison at the end makes it simpler to read. 

With lodash I was thinking that `_.forEach` instead of `for (let i=0; i<columns.length; i++)` might make the code read even more simply, but I don't think it's essential or a blocker. Sorry for not being more specific there!

Reply:
No worries! I think that since I'm traversing the array a bit more than just as an iterable, an explicit for loop might be best.